### PR TITLE
Cora experiment

### PR DESCRIPTION
**Pull request by nguyenhongson1996**
# Changes

- [x] Cora dataset
- [x] Model for Cora dataset
- [x] Training pipeline
- [x] Eval pipeline
- [x] E2E testing

### CODE REVIEW

File: configs/cora.yaml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+neptune_prj: NEPTUNE_PROJECT
+neptune_api_token: NEPTUNE_API_TOKEN
+ems_secret_name: NEPTUNE_SECRET_NAME
+experiment_name: CORA-GNN-RP
+seed: 1111
+is_train: true
+output_dir: ./assets/outputs
+checkpoint_path: null
+data_config:
+  type: CORA
```
Comment (hoangthanh283) [inline configs/cora.yaml:11]: We should add split type here as well

File: gnn/data_generator/datasets/planetoid_dataset.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        table = tabulate(table_data,
+                         headers=[],
+                         tablefmt='simple')
+        return '\n'.join(' ' * indent + line for line in table.splitlines())
+
+    def get_input_dim(self) -> int:
+        return self.x.shape[-1]
+
+    def get_num_classes(self) -> int:
+        return len(torch.unique(self.y))
```
Comment (hoangthanh283) [inline gnn/data_generator/datasets/planetoid_dataset.py]:
```suggestion
        return len((self.y.unique()))
```

Comment (hoangthanh283) [reply gnn/data_generator/datasets/planetoid_dataset.py]: But we can use dataset.num_classes directly

File: gnn/data_generator/datasets/planetoid_dataset.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+def get_adj_matrix(edge_index: torch.Tensor, num_nodes: int) -> torch.FloatTensor:
+    """
+    Construct dense ADJ matrix from edge indexes.
+    """
+    # Convert edge_indexes to scipy sparse matrix.
+    adj_matrix = sp.coo_matrix((np.ones(edge_index.shape[1]), (edge_index[0].numpy(), edge_index[1].numpy())),
+                               shape=(num_nodes, num_nodes), dtype=np.float32)
+    adj_matrix.setdiag(1)
+    return torch.FloatTensor(adj_matrix.todense())
```
Comment (hoangthanh283) [inline gnn/data_generator/datasets/planetoid_dataset.py:59]: We can directly use torch as well:
```python
values = torch.ones(edge_index.size(1))
sparse_adj_matrix = torch.sparse_coo_tensor(
    edge_index,
    values,
    size=(num_nodes, num_nodes)
)
dense_adj_matrix = sparse_adj_matrix.to_dense()
```

File: gnn/data_generator/datasets/planetoid_dataset.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+        dataset[0].val_mask.fill_(False)
+        dataset[0].val_mask[dataset[0].num_nodes - 1000:dataset[0].num_nodes - 500] = 1
+
+        dataset[0].test_mask.fill_(False)
+        dataset[0].test_mask[dataset[0].num_nodes - 500:] = 1
+    else:
+        dataset = Planetoid(path_to_save, dataset_name.value, split=split.value)
+
+    x = dataset[0].x if not normalize_features else T.NormalizeFeatures()
```
Comment (hoangthanh283) [inline gnn/data_generator/datasets/planetoid_dataset.py]:
```suggestion
    x = dataset[0].x
    if normalize_features:
       x = T.NormalizeFeatures()(x)
```
But I suggest to create an instance of normalization first at the begin so here we just call it without initialization

File: gnn/data_generator/datasets/planetoid_dataset.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    x = x.unsqueeze(0)
```
Comment (hoangthanh283) [inline gnn/data_generator/datasets/planetoid_dataset.py]:
```suggestion
    x = x.unsqueeze(0)  # For introducing the batch dim = 1.
```

File: gnn/trainer/training_procedures/planetoid_procedure.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+from torch import nn
+from tqdm import tqdm
+
+from gnn.data_generator.datasets.planetoid_dataset import PlanetoidDatasetName, get_planetoid_dataset
+from gnn.models.networks.deep_rp_robust_gcn import DeepRPRobustGCN
+from gnn.trainer.training_procedures.base_procedure import BaseProcedure
+from gnn.utils.metric_tracker import Dictlist
+
+
+class PlanetoidProcedure(BaseProcedure):
```
Comment (hoangthanh283) [inline gnn/trainer/training_procedures/planetoid_procedure.py:16]: We should add experiment tracking here to track the progress such as self.tb_writer and self.ems_exp.

File: gnn/trainer/training_procedures/planetoid_procedure.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+            self.global_step += 1
+            self.model.lambda_value = self.cosine_schedule_lambda(self.global_step, epoch,
+                                                                  self.config.num_epochs * num_epoch,
+                                                                  base_value=1e-4, max_value=1.0,
+                                                                  warmup_steps=5 * num_epoch)
+            table_data = [["Epoch", epoch]] + [[key, value] for key, value in tr_metric_scores.items()]
+            table = tabulate(table_data,
+                             headers=[],
+                             tablefmt='simple')
+            print(table)
```
Comment (hoangthanh283) [inline gnn/trainer/training_procedures/planetoid_procedure.py]: Use self.logger instead

File: scripts/demo_training_cora.py [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+EXP_POST_FIX = "DeepRPRobustGCN-rp-layer-at-last-cosine-scheduling"
+
+dataset = get_planetoid_dataset(PlanetoidDatasetName.CORA)
+
+cora_dataset = dataset
+
+rp_size = None
+lambda_value = 1  # 0.01  # 0.005
+model = DeepRPRobustGCN(input_dim=cora_dataset.get_input_dim(), output_dim=cora_dataset.get_num_classes(), num_edges=1,
```
Comment (hoangthanh283) [inline scripts/demo_training_cora.py]: Let's make the method as @propety so we only need to call:
```python
cora_dataset.input_dim
```

Comment (hoangthanh283) [reply scripts/demo_training_cora.py]: Same for the `cora_dataset.get_num_classes()`

File: scripts/demo_training_cora.py [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+                        lambda_value=lambda_value)
+
+# Write experiment names.
+config = GNNLearningWarper._from_config(args.config)
+procedure = PlanetoidProcedure(model, config)
```
Comment (hoangthanh283) [inline scripts/demo_training_cora.py]: Let's overwrite the experiment name according to the dataset name, particular config we use so we can track their experiments easier.
Such as:
```python
config.experiment_name = f"{config.experiment_name}-{PlanetoidDatasetName.CORA.value}-{EXP_POST_FIX}"

**Comment by hoangthanh283** [comment]
Let's add the experiment graph links here when we finish experiments (such as your Neptune links, so we can track, analyze reports later).

File: configs/cora.yaml (hunk 1 above)
Comment (nguyenhongson1996) [reply configs/cora.yaml:11]: We use all train/test/validation in our experiment so we do not need to specify.

File: gnn/data_generator/datasets/planetoid_dataset.py [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+    """
+    # Convert edge_indexes to scipy sparse matrix.
+    adj_matrix = sp.coo_matrix((np.ones(edge_index.shape[1]), (edge_index[0].numpy(), edge_index[1].numpy())),
+                               shape=(num_nodes, num_nodes), dtype=np.float32)
+    adj_matrix.setdiag(1)
+    return torch.FloatTensor(adj_matrix.todense())
+
+
+def get_planetoid_dataset(dataset_name: PlanetoidDatasetName, normalize_features=False,
+                          split=Split.PUBLIC) -> DatapileAdaptedPlanetoid:
```
Comment (hoangthanh283) [inline gnn/data_generator/datasets/planetoid_dataset.py:63]: Based on our previous discussion, we should use split="full" to use all nodes for training except ones in validation and test.
Ref: https://pytorch-geometric.readthedocs.io/en/2.5.3/generated/torch_geometric.datasets.Planetoid.html#torch_geometric.datasets.Planetoid

File: scripts/demo_training_cora.py [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+
+EXP_POST_FIX = "Planetoid-simple"
+
+dataset = get_planetoid_dataset(PlanetoidDatasetName.CORA)
+
+cora_dataset = dataset
+
+rp_size = None
+lambda_value = 1  # 0.01  # 0.005
+model = DeepRPPlanetoidGCN(input_dim=cora_dataset.get_input_dim(), output_dim=cora_dataset.num_classes,
```
Comment (hoangthanh283) [inline scripts/demo_training_cora.py:20]:
```suggestion
model = DeepRPPlanetoidGCN(input_dim=dataset.get_input_dim(), output_dim=dataset.num_classes,
```

File: scripts/demo_training_cora.py [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+
+parser = argparse.ArgumentParser(description="CL configurations")
+parser.add_argument("--config", default=None, type=str, help="Path to the configuration file.")
+args = parser.parse_args()
+
+EXP_POST_FIX = "Planetoid-simple"
+
+dataset = get_planetoid_dataset(PlanetoidDatasetName.CORA)
+
+cora_dataset = dataset
```
Comment (hoangthanh283) [inline scripts/demo_training_cora.py:16]:
```suggestion
```

**Review by hoangthanh283 — APPROVED** [review]
LGTM
